README.md

```diff
@@ -1,2 +1,159 @@
+programs written in other programming languages. 
+
+For instance, imagine that you are writing a AWS lambda runtime in Bash which doesn't do
+anything else other than echoing back whatever was sent through the input payload:
+
+```bash
+#!/bin/sh
+
+set -euo pipefail
+
+# function handler
+function handler () {
+  EVENT_DATA=$1
+  echo "$EVENT_DATA" 1>&2;
+  RESPONSE="Echoing request: '$EVENT_DATA'"
+
+  echo $RESPONSE
+}
+
+# Processing
+while true
+do
+  HEADERS="$(mktemp)"
+  # Get an event. The HTTP request will block until one is received
+  EVENT_DATA=$(curl -sS -LD "$HEADERS" -X GET "http://${AWS_LAMBDA_RUNTIME_API}/2018-06-01/runtime/invocation/next")
+
+  # Extract request ID by scraping response headers received above
+  REQUEST_ID=$(grep -Fi Lambda-Runtime-Aws-Request-Id "$HEADERS" | tr -d '[:space:]' | cut -d: -f2)
+
+  # Run the handler function from the script
+  RESPONSE=$($(echo "$_HANDLER" | cut -d. -f2) "$EVENT_DATA")
+
+  # Send the response
+  curl -X POST "http://${AWS_LAMBDA_RUNTIME_API}/2018-06-01/runtime/invocation/$REQUEST_ID/response"  -d "$RESPONSE"
+done
+```
+
+PS.: if we ignore shell built-ins such as `set` and `echo`, there are 8 processes being launched in which all 
+those OS tasks are performed for each of them.
+
+The execution time of this shell script is around **180-230ms on average**.
+
+However, if I, functionality-wise. implement, the exact same script but in C in which I compile it all in
+a single binary, I can get the execution time to be around **5-13ms on average**. 
+
+Creating this `'monolith-binary'` is a well-known technique that embedded developers tend to use. It saves both storage
+and memory space for low-resource devices. (Busybox and Toybox are well-known libraries that implement such approach).
+
+While I'm not aiming this idea for low-resource devices, it also happens to be a way to reduce execution time of a
+program (by eliminating most of the housekeeping OS tasks I mentioned earlier).
+
+
+# proposed solution
+This idea aims for reducing the execution times disparity between bash script and monolith-binaries for high-resource
+devices such as webservers and HPC clusters. It has 2 big components to make it work as shown below:
+
+## kernel space - new syscalls
+
+Currently, syscall `execve(2)` initiates a binary execution in the kernel. 
+
+From the [docs](https://man7.org/linux/man-pages/man2/execve.2.html):
+> execve() executes the program referred to by pathname.  This <br/>
+> causes the program that is currently being run by the calling <br/>
+> process to be replaced with a new program, with newly initialized <br/>
+> stack, heap, and (initialized and uninitialized) data segments.
+
+This triggers a bunch of userspace <-> kernel interactions that eventually
+run the program passed in as a parameter. The problem with that is:
+
+1. It's an I/O blocking operation
+2. Once triggered the program will run, there is no way back. That's equivalent
```

Comment:
Very noob question - is there a way to achieve this without requiring hacking on the kernel? E.g., that seems like the final implementation step, but perhaps there is a way to mock / test this before diving into that?

Reply:
That's a good question. As of now, I can't think of a way of testing it without fiddling with the kernel but I will give it some thought today... I agree with you that **if there is** an easier way to at least test the approach, we should pursue it.

Reply:
@vsoch  Happy new year!

I spent sometime thinking about a way to mock the pre-warm approach in userspace only and I think I got something.

Taking the following test script as an example:

```bash
#!/bin/bash

echo "test" > /tmp/file.txt
cat /tmp/file.txt
rm -f /tmp/file.txt
```

Then, we should implement:

- Compile some of the GNU coretools statically [1] (cat, echo, rm).
- Create a main program that will load this files from disk into a stack-allocated memory [2]. This program will be executed in two different ways: 
    

    - [Mode 1 - sync loading]: Each CLI utility is loaded into memory and executed when needed only  (like it happens in bash today)
    - [Mode 2 - pre-warm loading]: CLI utilties are loaded at the beginning of the main program (most likely in parallel - limited to the number of CPUs and Disk IO threads available). Those will remain in memory and be executed when needed. [3]

[1]: Dynamic Linker is kind of a hard one to emulate in userspace but most importantly, it would be adding an unnecessary complexity at this point. Additionally, we will have to modify those CLI utilities so that they don't call the `exit` syscall 😬 
[2]: Stack-allocated memory should be used instead of `malloc`/`brk` syscall to avoid skewing the results while running the benchmark as their latency varies a lot given due to memory utilisation/fragmentation at a given point in time.
[3]: Since we are not using the `execve` syscall, we will be "launching" programs by using the assembly `jmp` instruction instead. All we need to do is parse the ELF header and calculate the offset to the `_start` entrypoint.

## Additional observations:

### Linux Cache

To reduce outliers, we will drop page caches / slob objects (including dentries and inodes cache). If we don't do it, it would be really hard for us to know why results aren't consistent. Example of FS cache:

```
# clean fs cache
$ echo 3 | sudo tee /proc/sys/vm/drop_caches && sync && time -p git status
real 2,57
# warmed fs cache
$ time -p git status
real 0,40
```

So our tests should look something like this:

```bash
echo 0 > /proc/sys/kernel/nmi_watchdog
perf stat -r 100 -B bash -c "echo 3 > /proc/sys/vm/drop_caches && sync && ./main-program"
echo 1 > /proc/sys/kernel/nmi_watchdog
```

### Disable address space randomization

we may need to disable Address space layout randomization (ASLR) otherwise we might not be able to `jmp` into those CLI utilities to circumvent the `execve` syscall.

```bash
echo 0 | sudo tee /proc/sys/kernel/randomize_va_space
```

@vsoch How does that sound to you? Do you agree with that approach or do you suggest we do something differently?